### DISCUSSION TITLE
Language files - Move to Transiflex or Crowdin

### DISCUSSION THREAD

**Discussion by bovirus** [arendst/Tasmota#13328] [Ideas]
@ascillato 

Please consider for a better mangement of language files to move language files

https://github.com/arendst/Tasmota/tree/development/tasmota/language

to some host specifc for translation like Transiflex (suggested) or Crowdin.

In these hotsts the user could enable notification on language changes and get imemdiately a message when translation require changes.

Thanks.

**Comment by ascillato** [comment]
Thanks for the suggestion. We will check it.

**Reply by barbudor** [reply]
@bovirus What is your experience with Crowdin/Transiflex with an open-source project ?
We need to be able to support PR for code which also define new localisable strings.
It's easy for me to push a PR which define new text string.
But how would that work with Crowdin/Transiflex ?

Note that I have already used Crowdin to submit translations to a commercial softwares when new strings were added by the devs of the company.

If you can suggest a workflow that works with PR may be that could help adoption of such solution.

Both Crowdin and Transiflex have free plans for open-source projects:
https://crowdin.com/page/open-source-project-setup-request
https://opensource.com/article/17/6/open-source-localization-tools

**Comment by bovirus** [comment]
@barbudor

I didn't have a directly experience (I didn't have a repository on github).

Together with author of other sofwtare projects

** Password Safe

- Transiflex - https://www.transifex.com/passwordsafe/passwordsafe/dashboard/
- Github - https://github.com/pwsafe/pwsafe

** YDL-UI
- Transiflex - https://www.transifex.com/ydl-apps/ydl-ui/dashboard/
- Github - https://github.com/Maxstupo/ydl-ui
- 
** Network Manager
- Transiflex - https://www.transifex.com/BornToBeRoot/NETworkManager/dashboard/
- Github - https://github.com/BornToBeRoot/NETworkManager

we moved the language files from original github repository to Transiflex.

**Reply by barbudor** [reply]
Interresting that Transiflex was chosen the 3 times.
Seems like the open-source constraints for Crowdin are less acceptable from my point of view.

NetworkManager doc has a _how to add translation_ section here: https://github.com/BornToBeRoot/NETworkManager/blob/master/CONTRIBUTING.md

We could contact those guys. @ascillato would you prefer to contact them or are you happy that I do ?

**Reply by ascillato** [reply]
Great Idea. Go ahead to know how this is done and let's propose it to Theo.

**Comment by bovirus** [comment]
@barbudor

There are other projects that use Crowdin.

I propose 3 samples with Transiflex, because for me Transiflex has a better management of notification and translation advise for translators.

**Reply by barbudor** [reply]
@bovirus 
I can't seem to be able to get in touch with any of the 3 project owners above.
Do you have a direct contact with any of them ?
Thanks

**Reply by bovirus** [reply]
I will check the email refence.

I believe that you can start with BornToBeRoot,.

**Reply by barbudor** [reply]
I tried borntoberoot @ borntoberoot.net but didn't worked
No email neither on its website
Only a twitter and I don't use twitter

**Reply by bovirus** [reply]
I already wrote to BornToBeRoot o Transiflex.
Please wai an answer.

**Reply by BornToBeRoot** [reply]
Hi @bovirus @barbudor @ascillato 

I can try to help you. What information do you need exactly?

**Reply by barbudor** [reply]
Hi @BornToBeRoot 
Thanks for joining
We would be interrested to understand how you handle PR from occasional contributors (not collaborators) which include some new localizable strings.

For exemple, Tasmota gets a lot of contribution from people which are not neceseraly recurrent contributors. 
As of today, all localized strings are C/C++ macros #defined in a dediced per-language header files (en_uk.ino, fr_fr.ino, ...). Based on some compile time option and #ifdefs, the right header is used when building a binary for a given language (remember that Tasmota is an embedded firmware for MCU with limited capabilities and no possibility for run-time/dynamic language change. everything is statically linked when building the binary firmware).
So when someone submit a PR with new GUI or message texts, his PR includes changes in the "en_uk.ino" file to bring the new strings into the code. Those new strings are generally initially dupplicated as-is (english) to other language file, until anohter volunteer provide a PR with updated translations for his language.
This have worked well so far and allow anyone submitting a PR to provide new strings.

As @bovirus was suggesting more modern tools I was wondering how this could fit with the Tasmota contribution flow.
I understand that the submission of new translations (on the Transiflex platform) for existing strings would trigger a github action to get the newly translated strings commited into the code base (I've seen the history on your repo with such commits from a bot). But what I don't understand is how a contributor can submit a PR which include new code + the associated strings in the way they can later on be translated in the Web platform.
Should the contributor go to the Transiflex platform to create the strings or can they be part of their PR in some way ?

I hope I made myself clear.

Thanks you for sharing your experience.

**Reply by BornToBeRoot** [reply]
Hi,

so my worklflow is setup like this:

1) I have a file [Strings.resx](https://github.com/BornToBeRoot/NETworkManager/blob/master/Source/NETworkManager.Localization/Resources/Strings.resx) with my strings in english. This is similar to your `en_uk.ino` file. I only update this file when a new string is added to the application. You need to look at the transifex documentation if your file is supported.
2) Transifex can be setup to automatically look for changes in this file:
![image](https://user-images.githubusercontent.com/16019165/138518363-a8f9bdec-3345-42f7-a3db-e4cdbbd32c20.png)
The URL has to be the [RAW-file](https://raw.githubusercontent.com/BornToBeRoot/NETworkManager/master/Source/NETworkManager.Localization/Resources/Strings.resx)

or via the Transifex GitHub integration.

So when a PR is merged into master, it can take some time (max. 24h) until the string is available in transifex for translation.

3) Then your team/translators can translate (and review) the string
![image](https://user-images.githubusercontent.com/16019165/138518655-85529644-b16a-4734-b980-7cc63a698ebc.png)

4) Now you have 2 options

4.1) You can download all translations as zip (you get an email with the download link) and then copy the files with the translation in your repo

4.2) You can setup the [transifex integration](https://github.com/apps/transifex-integration)
![image](https://user-images.githubusercontent.com/16019165/138518951-9c54dbc1-ff76-4054-b4b6-c9081d79cc8e.png)

This is my config file. You can also re-map the language code to your format.
```
filters:
  - filter_type: file
    # all supported i18n types: https://docs.transifex.com/formats
    file_format: RESX
    source_language: en_US
    source_file: Source/NETworkManager.Localization/Resources/Strings.resx
    # path expression to translation files, must contain <lang> placeholder
    translation_files_expression: 'Source/NETworkManager.Localization/Resources/Strings.<lang>.resx'
settings:
    language_mapping:        
        zh_CN: zh-CN
        zh_TW: th-TW
        cs_CZ: cs-CZ
        etc.
```

Once a translation becomes 100% translated (or reviewed) a PR is send to your repository.

You can also manually send PRs when you plan a release and also want translations who are not completed yet.

![image](https://user-images.githubusercontent.com/16019165/138519433-c7d02c26-6c6c-4c79-917d-833e7244392d.png)

I have setup [mergify](https://mergify.io/) to merge this PRs when it comes from transifex and the CI is happy. Transifex will create 1 PR for each language...

```
pull_request_rules:
  - name: automatic merge for Transifex pull requests
    conditions:
      - author=transifex-integration[bot]
      - check-success=continuous-integration/appveyor/pr
    actions:
      merge:
        method: merge
```

I hope this helps 😄 

Greetings

**Reply by barbudor** [reply]
Thanks a lot @BornToBeRoot👍
This helps me a lot. I can start from here with a better understanding.

**Comment by Jason2866** [comment]
~~Show stopper 70$ per month for Transifex.~~
Edit: True free for Open Source Projects. It is well hidden! 
https://www.transifex.com/signup/open-source/

**Reply by bovirus** [reply]
@Jason2866 

Sorry but I didn't udnerstand.
Transiflex It's free for free sofwtare.

**Reply by comradekingu** [reply]
No it isn't. Read all the terms and services and what it actually says.
> for free to Open Source projects **that have no funding, revenue, and/or commercialization model**

For something to flaunt itself as "The localization platform with an Open Source DNA" actually being non-free software tells you all there is to know.

**Comment by barbudor** [comment]
@Jason2866 check the links above for the special plan for open source projects
Transiflex conditions seems easier than Crowndin

Sorry, sent before seing your correction

**Reply by comradekingu** [reply]
No they aren't. It is strictly for non-commercial open source projects.
At that, they will negotiate enterprise pricing when it befits them, effectively holding your community hostage to the only-as-a-service platform it is on. Use Weblate instead.
